src/Purpose.php

```diff
@@ -0,0 +1,140 @@
+<?php
+declare(strict_types=1);
+namespace ParagonIE\Paseto;
+
+use ParagonIE\Paseto\Keys\{
+    AsymmetricSecretKey,
+    AsymmetricPublicKey,
+    SymmetricKey
+};
+use ParagonIE\Paseto\Exception\InvalidPurposeException;
+
+final class Purpose
+{
+    const WHITELIST = [
+        'local',
+        'public',
+    ];
+
+    const EXPECTED_SENDING_KEYS = [
+        'local'  => SymmetricKey::class,
+        'public' => AsymmetricSecretKey::class,
+    ];
+
+    const EXPECTED_RECEIVING_KEYS = [
+        'local'  => SymmetricKey::class,
+        'public' => AsymmetricPublicKey::class,
+    ];
+
+    /** @var array<string, string> */
+    private static $sendingKeyToPurpose;
+
+    /** @var array<string, string> */
+    private static $receivingKeyToPurpose;
+
+    /**
+     * @var string
+     */
+    private $fuzz;
+
+    /**
+     * @var string
+     */
+    private $purpose;
+
+    /**
+     * @throws InvalidPurposeException
+     */
+    public function __construct(string $rawString)
+    {
+        if (!self::isValid($rawString)) {
+            throw new InvalidPurposeException('Unknown purpose: ' . $rawString);
+        }
+
+        $this->purpose = $rawString;
+        // prevent use of the == operator
+        // i.e. new Purpose('a') == new Purpose('a') will now be false
+        $this->fuzz = \random_bytes(16);
+    }
+
+    public static function local(): self
+    {
+        return new self('local');
+    }
+
+    public static function public(): self
+    {
+        return new self('public');
+    }
+
+    public function equals(self $purpose): bool
+    {
+        return \hash_equals($purpose->purpose, $this->purpose);
+    }
+
+    public function isSendingKeyValid(SendingKey $key): bool
```

Comment:
I should still probably add comments to all these methods, but this should be enough to gauge the general idea of the PR...